crates/bevy_ecs/src/error/bevy_error.rs

```diff
@@ -0,0 +1,268 @@
+use alloc::boxed::Box;
+use core::{
+    error::Error,
+    fmt::{Debug, Display},
+};
+
+/// The built in "universal" Bevy error type. This has a blanket [`From`] impl for any type that implements Rust's [`Error`],
+/// meaning it can be used as a "catch all" error.
+///
+/// # Backtraces
+///
+/// When used with the `backtrace` Cargo feature, it will capture a backtrace when the error is constructed (generally in the [`From`] impl]).
+/// When printed, the backtrace will be displayed. By default, the backtrace will be trimmed down to filter out noise. To see the full backtrace,
+/// set the `BEVY_BACKTRACE=full` environment variable.
+///
+/// # Usage
+///
+/// ```
+/// # use bevy_ecs::prelude::*;
+///
+/// fn fallible_system() -> Result<(), BevyError> {
+///     // This will result in Rust's built-in ParseIntError, which will automatically
+///     // be converted into a BevyError.
+///     let parsed: usize = "I am not a number".parse()?;
+///     Ok(())
+/// }
+/// ```
+pub struct BevyError {
+    inner: Box<dyn InnerError>,
```

Comment:
I haven't been following this too closely, but I wanted to point this out.

Something to think about is that `Box<dyn AnyTrait>` is a fat pointer (2 usizes)  where `Box<NormalType>` is thin (1 usize). For `Result<T, BevyError>` where T is small, this could bloat the size of the result. If looking up the error is rare, could we consider something like:

```suggestion
    inner: Box<Box<dyn InnerError>>,
```

That's two pointers, but cuts the size of the type in half. That could be meaningful if we're going to use this in a hot path. (Unless we'll be logging the error in the hot path too.)

Reply:
IMO, Errors are basically should never be the hot path.

Reply:
> The result size matters more than the double-indirection.
Just look at the size of `GltfError`

Reply:
Yeah double boxing is probably fine, but we could also avoid the double indirection by doing the vtable hackery that anyhow does.

But maybe we can defer that until later and just eat the double indirection for now (which is barely worth thinking about given that it is in the slow path)

Reply:
Hmm, if you double box the error, and put the backtrace in the *outer* box, then the inner box would just be `Box<dyn Error + ...>` and you could `impl<E> From<E> for BevyError where Box<dyn Error + ...>: From<E>` to get the `From<&str>` impl.  That might not work with the vtable tricks, though.

Reply:
> Hmm, if you double box the error, and put the backtrace in the outer box...

I actually think I did this trick in my short-lived custom error prototype.

Reply:
>  That might not work with the vtable tricks, though.

I think I'll take the `From<&str>` and "code simplicity" at the cost of an indirection in this case.

Reply:
Actually I do think it would be possible with the vtable tricks. So its probably just a "code aesthetics" tradeoff. For now, I'm willing to pay that price, but I'm not opposed to using the vtable in the future.